### ISSUE TITLE
github.com/tus/tusd/tree/master/vendor/golang.org/x/net/trace causes panic

### ISSUE THREAD

**Issue by rht** [tus/tusd#247]
This happens when I'm trying to import both tusd gcsstore and `"cloud.google.com/go/storage"`.
```
panic: http: multiple registrations for /debug/requests

goroutine 1 [running]:
net/http.(*ServeMux).Handle(0xed3a20, 0xa338bd, 0xf, 0xabca00, 0xa51c38)
        /nix/store/kq7cayifas0w60nac41l12mh3wq7h1cp-go-1.11.5/share/go/src/net/http/server.go:2377 +0x221
net/http.(*ServeMux).HandleFunc(0xed3a20, 0xa338bd, 0xf, 0xa51c38)
        /nix/store/kq7cayifas0w60nac41l12mh3wq7h1cp-go-1.11.5/share/go/src/net/http/server.go:2395 +0x5a
net/http.HandleFunc(0xa338bd, 0xf, 0xa51c38)
        /nix/store/kq7cayifas0w60nac41l12mh3wq7h1cp-go-1.11.5/share/go/src/net/http/server.go:2407 +0x4b
github.com/tus/tusd/vendor/golang.org/x/net/trace.init.0()
        $GOPATH/src/github.com/tus/tusd/vendor/golang.org/x/net/trace/trace.go:115 +0x42
```

Otherwise I had to manually patch this by doing `rm -rf $GOPATH/src/github.com/tus/tusd/vendor/golang.org/x/net/trace`.

See also https://github.com/etcd-io/etcd/issues/9357.

**Comment by Acconut** [comment]
Good catch, thanks for pointing this out. We need to find a way to move away from vendoring for the library but still have fixed dependency version for the tusd binary. I will see what we can do.

**Comment by Acconut** [comment]
I experimented with different dependency solution but unfortunately I couldn't get it to work properly. However, I will give this another shot when we start working on the next major release of tusd (https://github.com/tus/tusd/issues/212). Of course, help is always welcome!

**Comment by rht** [comment]
> I experimented with different dependency solution but unfortunately I couldn't get it to work properly. 

@Acconut  did you mean you excluded the option to use golang/dep according to https://github.com/grpc/grpc-go/issues/566#issuecomment-367861533 because it ensures only 1 version of `golang.org/x/net/trace` instead of having several pinned versions used independently?

**Comment by Acconut** [comment]
I didn't give every possible approach a shot as my time is also limited. But I explicitly decided to not try the dep tool for now as it does not seem to be very future-proof in contrast to the newer Go modules. I am not the only one concerned about this as you can see by the public feedback (https://github.com/golang/dep/issues/1959#issuecomment-407947097, https://github.com/golang/dep/issues/1996, etc). Especially now that Go 1.12 came with a more stable `go mod` tool I would rather not adopt dep from scratch.

However, dep may help you with your problems since dep removes the vendor directories of your dependencies and therefore removes the duplicate golang.org/x/net/trace. See https://golang.github.io/dep/docs/FAQ.html#a-id-my-dependers-don-t-use-dep-yet-what-should-i-do-a-my-dependers-don-t-use-dep-yet-what-should-i-do:

> If your dependers are using dep, this is not a concern, as dep takes care of stripping out nested vendor directories.

Of course, if you want to see if dep solves tusd's problem, I am happy to help you!

**Comment by rht** [comment]
I tested with `go mod init ...` and then `go build`, and it works! Found that `$GOPATH/pkg/mod/github.com/tus/tusd@v0.0.0-hash1-hash2`'s `vendor/` contains only `vendor.json` and `manifest`. And in my `go.sum` file I see several versions of `golang.org/x/net` indicating that each dependency use a different version. Not sure how the multiple init is solved though.

I will update https://github.com/grpc/grpc-go/issues/566#issuecomment-367861533 to point people to this solution as well.
